pcsx2/DEV9/Win32/pcap_io_win32.cpp

```diff
@@ -0,0 +1,284 @@
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PrecompiledHeader.h"
+
+#include "../pcap_io.h"
+
+HMODULE hpcap = nullptr;
+
+#define LOAD_FUNCTION(name)                                  \
+	fp_##name = (fp_##name##_t)GetProcAddress(hpcap, #name); \
+	if (fp_##name == nullptr)                                \
+	{                                                        \
+		FreeLibrary(hpcap);                                  \
+		Console.Error("%s not found", #name);                \
+		hpcap = nullptr;                                     \
+		return false;                                        \
+	}
+
+#define FUNCTION_SHIM_0_ARG(retType, name) \
+	typedef retType (*fp_##name##_t)();    \
+	fp_##name##_t fp_##name;               \
+	retType name()                         \
+	{                                      \
+		return fp_##name();                \
+	}
+#define FUNCTION_SHIM_1_ARG(retType, name, type1) \
+	typedef retType (*fp_##name##_t)(type1);      \
+	fp_##name##_t fp_##name;                      \
+	retType name(type1 a1)                        \
+	{                                             \
+		return fp_##name(a1);                     \
+	}
+#define FUNCTION_SHIM_2_ARG(retType, name, type1, type2) \
+	typedef retType (*fp_##name##_t)(type1, type2);      \
+	fp_##name##_t fp_##name;                             \
+	retType name(type1 a1, type2 a2)                     \
+	{                                                    \
+		return fp_##name(a1, a2);                        \
+	}
+#define FUNCTION_SHIM_3_ARG(retType, name, type1, type2, type3) \
+	typedef retType (*fp_##name##_t)(type1, type2, type3);      \
+	fp_##name##_t fp_##name;                                    \
+	retType name(type1 a1, type2 a2, type3 a3)                  \
+	{                                                           \
+		return fp_##name(a1, a2, a3);                           \
+	}
+#define FUNCTION_SHIM_4_ARG(retType, name, type1, type2, type3, type4) \
+	typedef retType (*fp_##name##_t)(type1, type2, type3, type4);      \
+	fp_##name##_t fp_##name;                                           \
+	retType name(type1 a1, type2 a2, type3 a3, type4 a4)               \
+	{                                                                  \
+		return fp_##name(a1, a2, a3, a4);                              \
+	}
+#define FUNCTION_SHIM_5_ARG(retType, name, type1, type2, type3, type4, type5) \
+	typedef retType (*fp_##name##_t)(type1, type2, type3, type4, type5);      \
+	fp_##name##_t fp_##name;                                                  \
+	retType name(type1 a1, type2 a2, type3 a3, type4 a4, type5 a5)            \
+	{                                                                         \
+		return fp_##name(a1, a2, a3, a4, a5);                                 \
+	}
+#define FUNCTION_SHIM_6_ARG(retType, name, type1, type2, type3, type4, type5, type6) \
```

Comment:
Shouldn't you replace this define hellfest with C++ generics? That seems really hacky

Reply:
I'm not sure how best to convert this to templates? how would I go about that?

Reply:
Not sure you need a template actually, you can probably do just fine with an auto and a variadic function, here's an example: 
https://en.cppreference.com/w/cpp/utility/variadic

Reply:
There isn't really a way to do this without macros without repeating names multiple times (which is a great way to mess up)

Personal recommendation, to avoid repeating for the definitions and the `LOAD_FUNCTION` calls: Move the list of functions to a separate file, then `#include` it once with `FUNCTION_SHIM_5_ARG`, etc defined to what they currently are, and once with `FUNCTION_SHIM_5_ARG` defined to `LOAD_FUNCTION` instead of manually rewriting everything for LOAD_FUNCTION

For a concrete example, LLVM does this a lot:
Definition: https://github.com/llvm/llvm-project/blob/5f022ad6ed8dc23094bb50f83608d5c786a821a9/mlir/lib/Parser/TokenKinds.def
Usage: https://github.com/llvm/llvm-project/blob/5f022ad6ed8dc23094bb50f83608d5c786a821a9/mlir/lib/Parser/Lexer.cpp#L216-L217

Reply:
> 
> 
> Not sure you need a template actually, you can probably do just fine with an auto and a variadic function, here's an example:
> https://en.cppreference.com/w/cpp/utility/variadic
variadic functions can't be used to define other functions at compile time (which is what I'm using the FUNCTION_SHIM_*_ARG macro for)

I could spit the FUNCTION_SHIM_*_ARG into smaller variadic macros, leaving FUNCTION_SHIM_*_ARG to reuse those macros, but I couldn't figure out how to make FUNCTION_SHIM_*_ARG fully variadic.

Let me know if you prefer to spit up the FUNCTION_SHIM_*_ARG macors or not?

> 
> 
> Personal recommendation, to avoid repeating for the definitions and the `LOAD_FUNCTION` calls: Move the list of functions to a separate file, then `#include` it once with `FUNCTION_SHIM_5_ARG`, etc defined to what they currently are, and once with `FUNCTION_SHIM_5_ARG` defined to `LOAD_FUNCTION` instead of manually rewriting everything for LOAD_FUNCTION

Thats a decent Idea.

Reply:
@TheLastRar What is your current implementation? I'll take a stab at it and see if I have a better idea, otherwise tellow's idea works too

Reply:
FUNCTION_SHIM_*_ARG defines a type, variable (holding the fp) and function (matching the signature of the function defined in the pcap headers). The function just calls the fp held in the variable the macro also defined

Take ``FUNCTION_SHIM_3_ARG(int, pcap_next_ex, pcap_t*, struct pcap_pkthdr**, const u_char**)``
It expands to;
```
typedef int (*fp_pcap_next_ex_t)(pcap_t*, struct pcap_pkthdr**, const u_char**);
pcap_next_ex_t fp_pcap_next_ex;
int pcap_next_ex(pcap_t* a1, struct pcap_pkthdr** a2, const u_char** a3)
{
    fp_pcap_next_ex(a1, a2, a3);
}
```

LOAD_FUNCTION takes the job of calling ``GetProcAddress()`` and assigning the returned function pointer to the correct variable.

Take ``LOAD_FUNCTION(pcap_next_ex);``
It expands to;
```
fp_pcap_next_ex = (fp_pcap_next_ex_t)GetProcAddress(hpcap, "pcap_next_ex");
if (fp_pcap_next_ex  == nullptr)
{
    FreeLibrary(hpcap);
    Console.Error("%s not found", "pcap_next_ex");
    hpcap = nullptr;
    return false;
}
```

Hopefully that is clear, let me know when you have attempted this

Reply:
To further clarify, ``pcap_next_ex()`` is defined in the pcap headers and used in ``pcap_io.cpp`` the ``FUNCTION_SHIM_*_ARG`` macro provides an implemtation for that already defined function, which then just calls the function pointer I load via ``LOAD_FUNCTION``